viewing_party/party.py

```diff
@@ -1,23 +1,180 @@
+    '''
+    friends_watched_list = get_friends_watched_list(user_data)
+    unique_watched_list = get_unique_list(friends_watched_list, user_data['watched'])
+    
+    unique_movies_list = []
+    movies_seen = set()
+
+    for movie in unique_watched_list:
+        movie_tuple = tuple(movie.items())
+        if movie_tuple not in movies_seen:
+            movies_seen.add(movie_tuple)
+            unique_movies_list.append(movie)
+
+    return unique_movies_list
+
+
+def get_friends_watched_list(user_data):
+    '''
+        Retrieves a list of all movies watched by the user's friends.
+    '''
+    return ([movie for friend in user_data['friends']
+            for movie in friend['watched']])
+
+def get_unique_list(source_list, exclusion_list):
+    '''
+    Returns a list of items found in the source list, excluding those 
+    from the exclusion list.
+    '''
+    return [item for item in source_list if item not in exclusion_list]
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    '''
+    Retrieves a list of recommended movies that friends have watched and
+    are available on the streaming services the user has access to.
+    '''
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+    return ([movie for movie in friends_unique_watched
+                if movie['host'] in user_data['subscriptions']])
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    '''
+    Compiles a list of recommended movies for the user. Movies added to
+    this list will match the user's most frequent genre, will have been
+    watched by at least one of the user's friends, and will not have been
+    watched yet by the user.
+    '''
+    if (not user_data['watched'] or not any(friend['watched']
+            for friend in user_data['friends'])):
+        return []
+
+    available_recs = get_available_recs(user_data)
+    most_watched_genre = get_most_watched_genre(user_data)
+
+    return ([rec for rec in available_recs
+                if rec['genre'] == most_watched_genre
+                and not any(movie['title'] == rec['title']
+                for movie in user_data['watched'])])
+
+def get_rec_from_favorites(user_data):
+    '''
+    Compiles a list of recommended movies from the user's favorites that
+    no friends have watched.
+    '''
+
+    if not user_data['favorites']:
+        return []
+    
+    if not user_data['friends']:
+        return user_data['favorites']
+
+    friends_watched_titles = ({
+        movie['title'] for friend in user_data['friends']
+        for movie in friend['watched']
+    })
+
+    rec_from_favorites = ([
+        favorite for favorite in user_data['favorites']
+        if favorite['title'] not in friends_watched_titles
+    ])

```

Comment:
This is starting to stretch how much we should put into comprehension. For me, this isn't as readable as it could be if we used a for loop.

What about something like 
```python
def get_rec_from_favorites(user_data):
    friends_movies = get_friends_watched_list(user_data)
    fav_movie_rec_list = []

    for movie in user_data["favorites"]:
        if movie not in friends_movies:
            fav_movie_rec_list.append(movie)

    return fav_movie_rec_list